CIP2/CIP2.md

```diff
@@ -0,0 +1,983 @@
+    * [Coin Selection in Cardano](#coin-selection-in-cardano)
+  * [Why is Coin Selection Non-Trivial?](#why-is-coin-selection-non-trivial)
+    * [The Transaction Size Limitation](#the-transaction-size-limitation)
+    * [The Problem of Dust](#the-problem-of-dust)
+    * [The Problem of Concurrency](#the-problem-of-concurrency)
+  * [Goals of Coin Selection Algorithms](#goals-of-coin-selection-algorithms)
+* [Definitions](#definitions)
+  * [Address](#address)
+  * [Coin Value](#coin-value)
+  * [Transaction](#transaction)
+  * [Transaction Input](#transaction-input)
+  * [Transaction Output](#transaction-output)
+  * [Spent Transaction Output](#spent-transaction-output)
+  * [Unspent Transaction Output](#unspent-transaction-output)
+  * [UTxO Set](#utxo-set)
+  * [Change Output](#change-output)
+  * [Dust Output](#dust-output)
+* [Interface](#interface)
+  * [Parameters](#parameters)
+    * [Requested Output Set](#requested-output-set)
+    * [Initial UTxO Set](#initial-utxo-set)
+    * [Maximum Input Count](#maximum-input-count)
+  * [Results](#results)
+    * [Coin Selection](#coin-selection)
+    * [Remaining UTxO Set](#remaining-utxo-set)
+  * [Properties](#properties)
+    * [Coverage of Payments](#coverage-of-payments)
+    * [Correctness of Change](#correctness-of-change)
+    * [Conservation of UTxO](#conservation-of-utxo)
+    * [Conservation of Outputs](#conservation-of-outputs)
+  * [Failure Modes](#failure-modes)
+    * [UTxO Balance Insufficient](#utxo-balance-insufficient)
+    * [UTxO Not Fragmented Enough](#utxo-not-fragmented-enough)
+    * [UTxO Fully Depleted](#utxo-fully-depleted)
+    * [Maximum Input Count Exceeded](#maximum-input-count-exceeded)
+* [Algorithms](#algorithms)
+  * [Largest-First](#largest-first)
+    * [State](#state)
+      * [Available UTxO List](#available-utxo-list)
+      * [Unpaid Output List](#unpaid-output-list)
+      * [Accumulated Coin Selection](#accumulated-coin-selection)
+    * [Computation](#computation)
+  * [Random-Improve](#random-improve)
+    * [Motivating Principles](#motivating-principles)
+      * [Principle 1: Dust Management](#principle-1-dust-management)
+      * [Principle 2: Change Management](#principle-2-change-management)
+      * [Principle 3: Performance Management](#principle-3-performance-management)
+    * [Cardinality](#cardinality-1)
+    * [State](#state-1)
+      * [Available UTxO Set](#available-utxo-set)
+      * [Accumulated Coin Selection](#accumulated-coin-selection-1)
+    * [Computation](#computation-1)
+      * [Phase 1: Random Selection](#phase-1-random-selection)
+      * [Phase 2: Improvement](#phase-2-improvement)
+    * [Termination](#termination-1)
+* [Reference Implementations](#reference-implementations)
+  * [Largest-First](#largest-first-1)
+  * [Random-Improve](#random-improve-1)
+* [External Resources](#external-resources)
+  * [Self Organisation in Coin Selection](#self-organisation-in-coin-selection)
+* [Copyright](#copyright)
+
+# Background
```

Comment:
Hi @KtorZ

> What about sticking to the CIP's naming convention and order of sections to make the structure more consistent with the template.

Naturally, there are some discrepancies between the structure of this document and the template defined by [CIP1](https://github.com/cardano-foundation/CIPs/blob/master/CIP1/CIP1.md), as this document was commissioned and written _before_ the CIP1 naming conventions became available.

I'm happy to make simple adjustments to make the structure more closely aligned, but am somewhat reluctant to adjust the section order _too much_, as this could entail extra work to restore the logical flow from one section to another. (The document was designed to be read in order from top to bottom.)

**Regarding:**
> Rename 'Background' to 'Motivation'

[CIP1](https://github.com/cardano-foundation/CIPs/blob/master/CIP1/CIP1.md#cip-format-and-structure) defines the Motivation section as:
> _The motivation is critical for CIPs that want to **change the Cardano protocol**. It should clearly explain why the existing protocol is inadequate to address the problem that the CIP solves._

I'm not sure that this is relevant to the [Background](https://github.com/jonathanknowles/CIPs/blob/CIP-JonathanKnowles-CoinSelectionAlgorithms/CIP2/CIP2.md#background) section, whose purpose is to give readers some _background understanding_ into what coin selection is. (I'm assuming a fairly diverse audience, where some readers will already be very familiar with coin selection, whereas others will need a bit of background explanation.)

My understanding of the purpose of a "Motivation" section is that it should answer the following questions:

1. Why this document was written.
2. Why readers should want to read it.

Do you agree? If so, I can add such a section. 

**Regarding**:
> Move 'Properties' and 'Motivating Principles' under a new 'Rationale' section to gives the concrete reason that motivated this design and also, the justification for its correctness.

After reading through [CIP1](https://github.com/cardano-foundation/CIPs/blob/master/CIP1/CIP1.md), I'm not yet clear of the difference between the [Motivation](https://github.com/cardano-foundation/CIPs/blob/master/CIP1/CIP1.md#cip-format-and-structure) and [Rationale](https://github.com/cardano-foundation/CIPs/blob/master/CIP1/CIP1.md#cip-format-and-structure) sections. They _seem_ to be almost synonymous to one another. Can you clarify the difference?

**Regarding**:
> Move 'Properties' ... under a new 'Rationale' section 

By _properties_, could you confirm that you're referring to [this section](https://github.com/jonathanknowles/CIPs/blob/CIP-JonathanKnowles-CoinSelectionAlgorithms/CIP2/CIP2.md#properties)?

Thanks!

Reply:
I've added a `Motivation` section to capture the motivation for writing this document.

Would be very grateful for any feedback on how this section might be improved!

See: 617c0deecbcafd4fdeff9891c24a17ce7ee7e662.